omnigibson/action_primitives/starter_semantic_action_primitives.py

```diff
@@ -0,0 +1,1898 @@
+        assert config["motor_type"] == "position", "Camera controller must be in position control mode"
+        use_delta = config["use_delta_commands"]
+
+        if use_delta:
+            cur_head_q = self.robot.get_joint_positions()[self.robot.camera_control_idx]
+            head_action = head_q - cur_head_q
+        else:
+            head_action = head_q
+        action[head_idx] = head_action
+        return action
+
+    def _get_head_goal_q(self, target_obj_pose):
+        """
+        Get goal joint positions for head to look at an object of interest,
+        If the object cannot be seen, return the current head joint positions.
+        """
+
+        # get current head joint positions
+        head1_joint = self.robot.joints["head_1_joint"]
+        head2_joint = self.robot.joints["head_2_joint"]
+        head1_joint_limits = [head1_joint.lower_limit, head1_joint.upper_limit]
+        head2_joint_limits = [head2_joint.lower_limit, head2_joint.upper_limit]
+        head1_joint_goal = head1_joint.get_state()[0][0]
+        head2_joint_goal = head2_joint.get_state()[0][0]
+
+        # grab robot and object poses
+        robot_pose = self.robot.get_position_orientation()
+        # obj_pose = obj.get_position_orientation()
+        obj_in_base = T.relative_pose_transform(*target_obj_pose, *robot_pose)
+
+        # compute angle between base and object in xy plane (parallel to floor)
+        theta = np.arctan2(obj_in_base[0][1], obj_in_base[0][0])
+        
+        # if it is possible to get object in view, compute both head joint positions
+        if head1_joint_limits[0] < theta < head1_joint_limits[1]:
+            head1_joint_goal = theta
+            
+            # compute angle between base and object in xz plane (perpendicular to floor)
+            head2_pose = self.robot.links["head_2_link"].get_position_orientation()
+            head2_in_base = T.relative_pose_transform(*head2_pose, *robot_pose)
+
+            phi = np.arctan2(obj_in_base[0][2] - head2_in_base[0][2], obj_in_base[0][0])
+            if head2_joint_limits[0] < phi < head2_joint_limits[1]:
+                head2_joint_goal = phi
+
+        # if not possible to look at object, return default head joint positions
+        else:
+            default_head_pos = self._get_reset_joint_pos()[self.robot.controller_action_idx["camera"]]
+            head1_joint_goal = default_head_pos[0]
+            head2_joint_goal = default_head_pos[1]
+
+        return [head1_joint_goal, head2_joint_goal]
+        
+    def _empty_action(self, arm_pose_to_keep=None):
+        """
+        No op action
+
+        Args:
+            arm_pose_to_keep (array) : if provided, returns action that keeps the arm in this pose.
+                Currently only used in IK control pose_absolute_ori mode: assumes [eef position] 
+        Returns:
+            np.array or None: Action array for one step for the robot to do nothing
+        """
```

Comment:
There turned out to be a good reason (inputting 0 action meant repeatedly re-solving IK which would cause drift) but I have patched this by making IKController not solve IK when given a small-enough delta.